### DISCUSSION TITLE
Recover basic configuration from SparkMax and SparkFlex

### DISCUSSION THREAD

**Discussion by Ramtech59Sensei** [wpilibsuite/2025Beta#40] [General]
There is no way to recover basic configuration from SparkMax and SparkFlex. The new library for 2025 add the capabilities to recover  
AbsoluteEncoderConfig , EncoderConfig, ect , but you need to still recover every basic parameters (like the IdleMode, Inverted and all other than are not grouped in classes ) from it getter and set it in a new configuration before applying to new controller. 

Will be nice to get the configuration from the controller  like:
SparkMaxConfig myconfig =  new SparkMaxConfig(mymotor);
then be able to change the parameters wanted to change or have a dedicated BasicConfig class than can be access the configAccessor and return
a the instance of the controller's  BasicConfig 

from this hierarchy:

    SparkMaxConfig
        inverted()
        idleMode()
        ...
        AbsoluteEncoderConfig
            inverted()
            ...
        EncoderConfig
            inverted()
            ...

To something like this?

    SparkMaxConfig
        BasicConfig
            inverted()
            idleMode()
            ...
        AbsoluteEncoderConfig
            inverted()
            ...
        EncoderConfig
            inverted()
            ...

**Comment by Ramtech59Sensei** [comment]
I notice non of the config class have a getter. This mean there is no way to copy it with recover every value one by one , by adding a getter we can copy the whole config class at once 
Also I can't find in the code a way to retrieve the all config classes from one controller  example SparkMaxConfig, there is ways to apply one but not to get it

**Reply by jfabellera** [reply]
The reason there are no getters is because the config classes are not tied to any single device and are not meant to get the config parameter from the devices. They are only for setting. Instead you would use the `configAccessor` member (see https://github.com/wpilibsuite/2025Beta/discussions/40#discussioncomment-11245008).

If you want to copy a config, you can use [`apply()`](https://codedocs.revrobotics.com/java/com/revrobotics/spark/config/sparkbaseconfig#apply(com.revrobotics.spark.config.SparkBaseConfig)). I linked to one of `SparkBaseConfig`'s apply methods, but there is one for each config class, and multiple for configs with nested configs.

>  This mean there is no way to copy it with recover every value one by one

I don't understand what you mean by this. Could you explain further?

**Comment by jfabellera** [comment]
I am going to convert this to a discussion because I don't think it addresses a code issue with REVLib, rather a question/suggestion for the library.

**Comment by jfabellera** [comment]
> Will be nice to get the configuration from the controller like:
> SparkMaxConfig myconfig = new SparkMaxConfig(mymotor);

What would be the benefits of doing it this way?

If the user wants to retrieve a parameter from the device, they can use the `configAccessor` member of the `SparkMax` or `SparkFlex` class, which queries the value directly from the device.

Config classes are set up such that they are only structures that allow the user to declaratively specify which parameters they want to be set. If a parameter is not specified in the config, they will not be modified on the device _**unless**_ `kResetSafeParameters` is passed when calling [`configure()`](https://codedocs.revrobotics.com/java/com/revrobotics/spark/sparkbase#configure(com.revrobotics.spark.config.SparkBaseConfig,com.revrobotics.spark.SparkBase.ResetMode,com.revrobotics.spark.SparkBase.PersistMode)), which sets most parameters to their defaults before setting the specified parameters.

If you want to configure the device from its current state, you would pass `kNoResetSafeParameters` when calling `configure()`. This would only set parameters that you specify without affecting the rest of the configuration. Doing it like this is more efficient than retrieving every single parameter from the device, modifying them, and sending them back to the device.

**Comment by jfabellera** [comment]
> To something like this?
> 
> ```
> SparkMaxConfig
>     BasicConfig
>         inverted()
>         idleMode()
>         ...
>     AbsoluteEncoderConfig
>         inverted()
>         ...
>     EncoderConfig
>         inverted()
>         ...
> ```

In terms of organization, I agree that something like this would be better. The reason we have some straggler configurations is because they didn't all fit into one single category.

However, what would be the benefits of doing it like this, other than for conceptual organization?

**Reply by Ramtech59Sensei** [reply]
So every time I create a new Config object is an empty object with no parameters in it and as I change parameters those parameter are add to the Config Object.
I understand now how this class config work, then organization is irrelevant also 
Now to be 100% sure 
If I want to change the position conversion factor of my sparkmax or SparkFlex 
I just n ed to create an object Config for the encoder, change position conversion factor and apply the config object with configure!
Correct!

**Reply by jfabellera** [reply]
Yes that is correct
